roxie/roxiemem/roxiemem.cpp

```diff
@@ -5081,30 +5077,31 @@ class CDataBufferManager : public CInterface, implements IDataBufferManager
         {
             if (curBlock)
             {
-                DataBufferBottom *bottom = (DataBufferBottom *) curBlock;
+                DataBufferBottom *bottom = curBlock;
+                assertex(((memsize_t)bottom & HEAP_PAGE_OFFSET_MASK) == 0);
                 CriticalBlock c(bottom->crit);
                 if (bottom->freeChain)
                 {
                     atomic_inc(&dataBuffersActive);
-                    HeapletBase::link(curBlock);
-                    DataBufferBase *x = bottom->freeChain;
+                    curBlock->Link();
+                    DataBuffer *x = bottom->freeChain;
                     bottom->freeChain = x->next;
                     x->next = NULL;
                     if (memTraceLevel >= 4)
                         DBGLOG("RoxieMemMgr: CDataBufferManager::allocate() reallocated DataBuffer - addr=%p", x);
                     return ::new(x) DataBuffer();
                 }
-                else if ((memsize_t)(nextAddr - curBlock) <= HEAP_ALIGNMENT_SIZE-DATA_ALIGNMENT_SIZE)
+                else if ((((memsize_t)nextAddr) & HEAP_PAGE_OFFSET_MASK) != 0) // Is there any space in the current block
```

Comment:
Yes.
It was filling up from DATA_ALIGNMENT_SIZE .. HEAP_ALIGNMENT_SIZE.  The previous test could have equally said

(memsize_t)(nextAddr - curBlock) < HEAP_ALIGNMENT_SIZE)

Because HEAP_ALIGNMENT_SIZE is a multiple of DATA_ALIGNMENT_SIZE .  (I will add a static assert to check that).

Stage 2 of this change is to move the DataBufferBottom out of the page, so I wanted to remove curBlock form the test.  (Having though a little more about that I will need to change this check again to something similar, so maybe I will do that now).